src/main/java/core/stack/MateStack.java

```diff
@@ -1,19 +1,53 @@
 package core.stack;
 
+import java.util.EmptyStackException;
+
 public class MateStack<T> {
-    public void push(T value) {
+    private Node<T> top;
+    private int size;
 
+    public MateStack() {
+        this.top = null;
+        this.size = 0;
+    }
+
+    public void push(T value) {
+        top = new Node<>(value, top);
+        size++;
     }
 
     public T peek() {
-        return null;
+        if (isEmpty()) {
+            throw new EmptyStackException();
+        }
+        return top.value;
     }
 
     public T pop() {
-        return null;
+        if (isEmpty()) {
+            throw new EmptyStackException();
+        }
+        T value = top.value;
+        top = top.next;
+        size--;
+        return value;
     }
 
     public int size() {
-        return 0;
+        return size;
+    }
+
+    private boolean isEmpty() {
+        return top == null;
+    }
```

Comment:
Consider making the isEmpty method public. While it's currently being used internally, making it public could be beneficial if other classes or users of the MateStack class need to check if the stack is empty without modifying the stack's state.